### PR TITLE
[azcore] move time parsing to core

### PR DESCRIPTION
**Pull request by gracewilcox**
Part of https://github.com/Azure/azure-sdk-for-go/issues/22314

Moving time to core. Will be consumed by generator.

### CODE REVIEW

File: sdk/azcore/runtime/datetime/RFC3339.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	utcDateTime        = "2006-01-02T15:04:05.999999999"
+	utcDateTimeJSON    = `"` + utcDateTime + `"`
+	utcDateTimeNoT     = "2006-01-02 15:04:05.999999999"
+	utcDateTimeJSONNoT = `"` + utcDateTimeNoT + `"`
+	dateTimeNoT        = `2006-01-02 15:04:05.999999999Z07:00`
+	dateTimeJSON       = `"` + time.RFC3339Nano + `"`
+	dateTimeJSONNoT    = `"` + dateTimeNoT + `"`
+)
+
+type DateTimeRFC3339 time.Time
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/RFC3339.go]: Remove `DateTime` prefix.

Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/RFC3339.go]: Also will need docs.

File: sdk/azcore/datetime/date_type.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+)
+
+const (
+	fullDateJSON = `2006-01-02`
+	jsonFormat   = `%04d-%02d-%02d`
+)
+
+type DateType time.Time
```
Comment (jhendrixMSFT) [inline sdk/azcore/datetime/date_type.go]: `type PlainDate time.Time`

Comment (gracewilcox) [reply sdk/azcore/datetime/date_type.go]: fixed

File: sdk/azcore/runtime/datetime/RFC3339.go (hunk 1 above)
Comment (gracewilcox) [reply sdk/azcore/runtime/datetime/RFC3339.go]: fixed

File: sdk/azcore/runtime/datetime/plain_time_test.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestPlainTime(t *testing.T) {
+	originalTime := time.Date(2023, time.June, 15, 10, 30, 45, 0, time.UTC)
+	pt := PlainTime(originalTime)
+	result := pt.String()
+	require.Equal(t, "10:30:45Z", result)
+
+	// jsonBytes, err := pt.MarshalJSON()
```
Comment (gracewilcox) [inline sdk/azcore/runtime/datetime/plain_time_test.go]: @jhendrixMSFT, would roundtrip on Marshal/UnmarshalJSON be expected to fail? Just copied over from the current implementation, and don't know if that's expected or if I need to adjust those methods

Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/plain_time_test.go]: It should work.  My guess is that the marshaled value isn't quoted.  Is that the case?

File: sdk/azcore/runtime/datetime/plain_date.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	"time"
+)
+
+// Time is a type constraint that represents the supported time types in the datetime package
+type Time interface {
+	PlainDate | PlainTime | RFC1123 | RFC3339 | Unix
+}
+
+const (
+	fullDateJSON = `2006-01-02`
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/plain_date.go]:
```go
plainDateJSON = `"2006-01-02"`
```

File: sdk/azcore/runtime/datetime/plain_date.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	jsonFormat   = `%04d-%02d-%02d`
+)
+
+// PlainDate represents a date value without time information in YYYY-MM-DD format.
+// It wraps time.Time and can be marshaled to and unmarshaled from JSON.
+type PlainDate time.Time
+
+// MarshalJSON marshals the PlainDate to a JSON byte slice.
+func (t PlainDate) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(jsonFormat, time.Time(t).Year(), time.Time(t).Month(), time.Time(t).Day())), nil
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/plain_date.go]: You can simplify this.
```go
return []byte(time.Time(t).Format(plainDateJSON)), nil
```

File: sdk/azcore/runtime/datetime/plain_date.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+package datetime
+
+import (
+	"fmt"
+	"time"
+)
+
+// Time is a type constraint that represents the supported time types in the datetime package
+type Time interface {
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/plain_date.go]: How about `TypeConstraint`?

File: sdk/azcore/runtime/datetime/plain_date_test.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestPlainDate(t *testing.T) {
+	originalDate := time.Date(2023, time.January, 15, 0, 0, 0, 0, time.UTC)
+	pd := PlainDate(originalDate)
+
+	jsonBytes, err := pd.MarshalJSON()
+	require.NoError(t, err)
+	require.Equal(t, "2023-01-15", string(jsonBytes))
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/plain_date_test.go]:
```go
require.Equal(t, `"2023-01-15"`, string(jsonBytes))
```

File: sdk/azcore/runtime/datetime/plain_date.go (hunk 6 above)
Comment (gracewilcox) [reply sdk/azcore/runtime/datetime/plain_date.go]: Down for either name! Could be datetime.TypeConstraint or datetime.Time (or similiar). I just leaned towards Time because the type constraint examples I found were usual named after the object (example: Integer, Number, etc https://pkg.go.dev/golang.org/x/exp/constraints)

Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/plain_date.go]: Ah interesting.  Maybe just `Constraints` then?

File: sdk/azcore/runtime/datetime/plain_time.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+package datetime
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+const (
+	utcTimeJSON    = `"15:04:05.999999999"`
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/plain_time.go]: I think we should name these `plainTime`, `plainTimeJSON` etc...

Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/plain_time.go]: We should also concatenate values instead of duplicating them (I got lazy).

File: sdk/azcore/runtime/datetime/plain_time.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	plainTimeJSON    = `"` + plainTime + `"`
+)
+
+// PlainTime represents a time value without date information. It supports HH:MM:SS format
+// with optional nanosecond precision and timezone information.
+type PlainTime time.Time
+
+// MarshalJSON marshals the PlainTime to a JSON byte slice.
+func (t PlainTime) MarshalJSON() ([]byte, error) {
+	s, _ := t.MarshalText()
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/plain_time.go]: We can simplify this.
```go
return []byte(time.Time(t).Format(plainTimeJSON)), nil
```

File: sdk/azcore/runtime/datetime/plain_time.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	}
+	layout := plainTimeUTC
+	if tzOffsetRegex.Match(data) {
+		layout = plainTime
+	}
+	return t.Parse(layout, string(data))
+}
+
+// Parse parses a time string using the specified layout
+func (t *PlainTime) Parse(layout, value string) error {
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/plain_time.go]: I don't recall where this is used in the generated code.  I only see it here and on `RFC3339`.  Can these be removed, or do we need them for _all_ time types?

File: sdk/azcore/runtime/datetime/plain_time.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+// MarshalJSON marshals the PlainTime to a JSON byte slice.
+func (t PlainTime) MarshalJSON() ([]byte, error) {
+	s, _ := t.MarshalText()
+	return []byte(fmt.Sprintf("\"%s\"", s)), nil
+}
+
+// MarshalText returns a textual representation of PlainTime
+func (t PlainTime) MarshalText() ([]byte, error) {
+	tt := time.Time(t)
+	return []byte(tt.Format(plainTime)), nil
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/plain_time.go]: I think this should be `plainTimeUTC` (same in the `String()` method).

File: sdk/azcore/CHANGELOG.md [hunk 12]
```diff
@@ -1,8 +1,10 @@
 # Release History
 
-## 1.20.1-beta.1 (Unreleased)
+## 1.30.0 (Unreleased)
```
Comment (jhendrixMSFT) [inline sdk/azcore/CHANGELOG.md]: Will also need to update the version `const`.

Comment (chlowell) [inline sdk/azcore/CHANGELOG.md]:
```suggestion
## 1.21.0 (Unreleased)
```

Comment (gracewilcox) [reply sdk/azcore/CHANGELOG.md]: Since I'm adding new public api, think it should be a minor version bump instead of a patch version

Comment (chlowell) [reply sdk/azcore/CHANGELOG.md]: Agreed, however the next minor version is 1.21.0, not 1.30.0. Current version is 1.20.0, next patch would be 1.20.1

File: sdk/azcore/runtime/datetime/plain_time.go (hunk 10 above)
Comment (gracewilcox) [reply sdk/azcore/runtime/datetime/plain_time.go]: don't think it's use in generated code. Think we can make internal

File: sdk/azcore/CHANGELOG.md (hunk 12 above)
Comment (gracewilcox) [reply sdk/azcore/CHANGELOG.md]: you're right, I can't count

**Comment by jhendrixMSFT** [comment]
We should port the time-specific tests from the emitter repo to here.
- https://github.com/Azure/autorest.go/pull/1129
- https://github.com/Azure/autorest.go/pull/1152

Both contain some corner-case tests for parsing time formats.

File: sdk/azcore/runtime/datetime/RFC3339.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+package datetime
+
+import (
+	"regexp"
+	"strings"
+	"time"
+)
+
+// Azure reports time in UTC but it doesn't include the 'Z' time zone suffix in some cases.
+var tzOffsetRegex = regexp.MustCompile(`(?:Z|z|\+|-)(?:\d+:\d+)*"*$`)
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/RFC3339.go:13]: The suggestion is wrong.  And while this might not be perfect, it's been working for years so I'd prefer to not change it.
